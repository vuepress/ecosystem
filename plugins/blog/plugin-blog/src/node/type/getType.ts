import { entries, fromEntries, removeLeadingSlash } from '@vuepress/helper'
import type { PageOptions } from 'vuepress/core'
import { colors } from 'vuepress/utils'
import type { TypeMap, TypesMap } from '../../shared/index.js'
import type { PagesMap } from '../getPagesMap.js'
import { logger } from '../logger.js'
import type { BlogTypeOptions } from '../options.js'
import type { Store } from '../store.js'
import { getPagePath } from '../utils.js'

export const getType = (
  pagesMap: PagesMap,
  store: Store,
  type: BlogTypeOptions[],
  slugify: (name: string) => string,
  isDebug = false,
): {
  typesMap: TypesMap
  pageOptions: PageOptions[]
} => {
  const result = type.map(
    ({
      key,
      sorter = (): number => -1,
      filter = (): boolean => true,
      path = '/:key/',
      layout = 'Layout',
      frontmatter = (): Record<string, string> => ({}),
    }) => {
      if (isDebug) logger.info(`Generating ${colors.cyan(key)} type.\n`)

      const pageOptions: PageOptions[] = []
      const typeMap: TypeMap = {}

      entries(pagesMap).forEach(([localePath, pages]) => {
        // get type page path
        const pagePath = path
          ? `${localePath}${removeLeadingSlash(
              path.replace(/:key/g, slugify(key)),
            )}`
          : ''

        // get type indexes
        const indexes = store.addItems(
          // get page paths
          pages
            .filter(filter)
            .sort(sorter)
            .map(({ path }) => path),
        )

        if (pagePath)
          pageOptions.push({
            path: pagePath,
            frontmatter: {
              ...frontmatter(localePath),
              blog: {
                type: 'type',
                key,
              },
              layout,
            },
          })

        if (isDebug)
          logger.info(
            `${key} type in locale ${localePath}: found ${indexes.length} items\n`,
          )

        typeMap[localePath] = { path: getPagePath(pagePath), indexes }
      })

      return { key, typeMap, pageOptions }
    },
  )

  return {
    typesMap: fromEntries(result.map(({ key, typeMap }) => [key, typeMap])),
    pageOptions: result.map(({ pageOptions }) => pageOptions).flat(),
  }
}
