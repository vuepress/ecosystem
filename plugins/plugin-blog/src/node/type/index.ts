export * from './getType.js'
export * from './getTypeOptions.js'
export * from './prepareTypesMap.js'
