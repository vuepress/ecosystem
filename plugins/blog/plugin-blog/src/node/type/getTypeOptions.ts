import { isString } from '@vuepress/helper'
import { colors } from 'vuepress/utils'
import { logger } from '../logger.js'
import type { BlogTypeOptions } from '../options.js'

export const getTypeOptions = (types: BlogTypeOptions[]): BlogTypeOptions[] =>
  types.filter((type, index) => {
    const { key } = type

    if (!isString(key) || !key) {
      logger.error(
        `Invalid ${colors.magenta('key')} option ${colors.cyan(
          key,
        )} in ${colors.cyan(`type[${index}]`)}`,
      )

      return false
    }

    return true
  })
