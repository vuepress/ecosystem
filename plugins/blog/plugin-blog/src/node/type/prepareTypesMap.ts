import type { App } from 'vuepress/core'
import type { TypesMap } from '../../shared/index.js'

const HMR_CODE = `
if (import.meta.webpackHot) {
  import.meta.webpackHot.accept();
  if (__VUE_HMR_RUNTIME__.updateBlogType)
    __VUE_HMR_RUNTIME__.updateBlogType(typesMap);
}

if (import.meta.hot)
  import.meta.hot.accept(({ typesMap }) => {
    __VUE_HMR_RUNTIME__.updateBlogType(typesMap);
  });
`

export const prepareTypesMap = async (
  app: App,
  typesMapData: TypesMap,
): Promise<void> => {
  await app.writeTemp(
    `blog/type.js`,
    `\
      export const typesMap = ${JSON.stringify(typesMapData)};
      ${app.env.isDev ? HMR_CODE : ''}
      `,
  )
}
