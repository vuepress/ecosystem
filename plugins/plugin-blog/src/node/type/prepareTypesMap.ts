import type { App } from 'vuepress/core'
import type { TypesMap } from '../../shared/index.js'

const HMR_CODE = `
if (import.meta.webpackHot) {
  import.meta.webpackHot.accept();
  if (__VUE_HMR_RUNTIME__.updateBlogType)
    __VUE_HMR_RUNTIME__.updateBlogType(typeMap);
}

if (import.meta.hot)
  import.meta.hot.accept(({ typeMap }) => {
    __VUE_HMR_RUNTIME__.updateBlogType(typeMap);
  });
`

export const prepareTypesMap = async (
  app: App,
  typesMapData: TypesMap,
): Promise<void> => {
  await app.writeTemp(
    `blog/type.js`,
    `\
      export const typesMap = ${JSON.stringify(typesMapData)};
      ${app.env.isDev ? HMR_CODE : ''}
      `,
  )
}
