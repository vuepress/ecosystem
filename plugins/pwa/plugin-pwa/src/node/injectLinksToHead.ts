import type { App } from 'vuepress/core'
import { isPlainObject } from 'vuepress/shared'
import type { PwaPluginOptions } from './options.js'

export const injectLinksToHead = (
  app: App,
  { favicon, manifest, themeColor = '#46bd87', apple }: PwaPluginOptions,
): void => {
  const { base, head } = app.siteData
  const metaKeys: string[] = []
  const linkKeys: string[] = []

  // Generate Hash for Head
  head.forEach((item) => {
    if (item[0] === 'meta') metaKeys.push(item[1].name as string)
    else if (item[0] === 'link') linkKeys.push(item[1].rel as string)
  })

  const appTitle = manifest?.name ?? app.siteData.title
  let fallBackIcon = ''

  const setLink = (
    rel: string,
    href: string,
    more: Record<string, string> = {},
  ): void => {
    if (!linkKeys.includes(rel)) head.push(['link', { rel, href, ...more }])
  }

  const setMeta = (
    name: string,
    content: string,
    more: Record<string, string> = {},
  ): void => {
    if (!metaKeys.includes(name))
      head.push(['meta', { name, content, ...more }])
  }

  if (appTitle) setMeta('application-name', appTitle)

  setMeta('mobile-web-app-capable', 'yes')
  setMeta('theme-color', themeColor || '#46bd87')

  // eslint-disable-next-line @typescript-eslint/no-deprecated
  if (isPlainObject(apple) && apple.statusBarColor)
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    setMeta('apple-mobile-web-app-status-bar-style', apple.statusBarColor)

  if (favicon) setLink('icon', favicon)

  setLink('manifest', `${base}manifest.webmanifest`, {
    crossorigin: 'use-credentials',
  })

  if (manifest?.icons) {
    const { icons } = manifest

    if (icons.length > 0) {
      fallBackIcon = icons[0].src

      manifest.icons.forEach((icon) => {
        if (icon.type)
          setLink('icon', icon.src, { type: icon.type, sizes: icon.sizes })
        else setLink('icon', icon.src, { sizes: icon.sizes })
      })
    }
  }

  if (isPlainObject(apple) && (apple.icon || fallBackIcon)) {
    setLink('apple-touch-icon', apple.icon || fallBackIcon)
    if (apple.maskIcon)
      setLink('mask-icon', apple.maskIcon, {
        color: themeColor || '#46bd87',
      })
  } else if (apple !== false && fallBackIcon) {
    setLink('apple-touch-icon', fallBackIcon)
  }
}
