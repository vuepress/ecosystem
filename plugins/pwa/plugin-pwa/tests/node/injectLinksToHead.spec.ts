import { describe, expect, it } from 'vitest'
import { createBaseApp } from 'vuepress/core'
import type { PwaPluginOptions } from '../../src/node/index.js'
import { injectLinksToHead } from '../../src/node/injectLinksToHead.js'

const options: PwaPluginOptions = {
  favicon: '/favicon.ico',

  manifest: {
    icons: [
      {
        src: '/assets/icon/chrome-192.png',
        sizes: '192x192',
      },
      {
        src: '/assets/icon/chrome-512.png',
        sizes: '512x512',
      },
      {
        src: '/assets/icon/chrome-mask-192.png',
        sizes: '192x192',
        purpose: 'maskable',
        type: 'image/png',
      },
      {
        src: '/assets/icon/chrome-mask-512.png',
        sizes: '512x512',
        purpose: 'maskable',
        type: 'image/png',
      },
    ],
    shortcuts: [
      {
        name: 'Guide',
        url: '/guide/',
      },
      {
        name: 'Config',
        url: '/config/',
      },
      {
        name: 'Basic',
        url: '/basic/',
      },
    ],
  },

  themeColor: '#46bd87',
  apple: {
    icon: '/img/icon/appleIcon152.png',
    maskIcon: '/icons/safari-pinned-tab.svg',
  },
}

describe('Test head function', () => {
  it('should generate PWA tags because they do not exist', () => {
    // @ts-expect-error: Fake app
    const app = createBaseApp({
      base: '/',
      title: 'VuePress',
      source: '',
    })

    expect(injectLinksToHead(app, options)).toMatchSnapshot()
  })

  it('should read manifest', () => {
    // @ts-expect-error: Fake app
    const app = createBaseApp({
      base: '/',
      title: 'VuePress',
      source: '',
    })

    const optionsWithManifest: PwaPluginOptions = {
      manifest: {
        icons: [
          {
            src: '/assets/icon/chrome-192.png',
            sizes: '192x192',
            type: 'image/png',
          },
        ],
      },
    }

    expect(injectLinksToHead(app, optionsWithManifest)).toMatchSnapshot()
  })

  it('should not generate tags if they exist', () => {
    // @ts-expect-error: Fake app
    const app = createBaseApp({
      base: '/',
      title: 'VuePress',
      head: [
        ['link', { rel: 'icon', href: '/icon.ico' }],
        ['meta', { name: 'theme-color', content: '#ffffff' }],
      ],
      source: '',
    })

    expect(injectLinksToHead(app, options)).toMatchSnapshot()
  })
})
