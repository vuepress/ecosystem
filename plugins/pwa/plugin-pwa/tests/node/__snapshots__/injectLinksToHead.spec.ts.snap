// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test head function > should generate PWA tags because they do not exist 1`] = `
[
  [
    "meta",
    {
      "content": "VuePress",
      "name": "application-name",
    },
  ],
  [
    "meta",
    {
      "content": "yes",
      "name": "mobile-web-app-capable",
    },
  ],
  [
    "meta",
    {
      "content": "#46bd87",
      "name": "theme-color",
    },
  ],
  [
    "link",
    {
      "href": "/favicon.ico",
      "rel": "icon",
    },
  ],
  [
    "link",
    {
      "crossorigin": "use-credentials",
      "href": "/manifest.webmanifest",
      "rel": "manifest",
    },
  ],
  [
    "link",
    {
      "href": "/assets/icon/chrome-192.png",
      "rel": "icon",
      "sizes": "192x192",
    },
  ],
  [
    "link",
    {
      "href": "/assets/icon/chrome-512.png",
      "rel": "icon",
      "sizes": "512x512",
    },
  ],
  [
    "link",
    {
      "href": "/assets/icon/chrome-mask-192.png",
      "rel": "icon",
      "sizes": "192x192",
      "type": "image/png",
    },
  ],
  [
    "link",
    {
      "href": "/assets/icon/chrome-mask-512.png",
      "rel": "icon",
      "sizes": "512x512",
      "type": "image/png",
    },
  ],
  [
    "link",
    {
      "href": "/img/icon/appleIcon152.png",
      "rel": "apple-touch-icon",
    },
  ],
  [
    "link",
    {
      "color": "#46bd87",
      "href": "/icons/safari-pinned-tab.svg",
      "rel": "mask-icon",
    },
  ],
]
`;

exports[`Test head function > should not generate tags if them exist 1`] = `
[
  [
    "link",
    {
      "href": "/icon.ico",
      "rel": "icon",
    },
  ],
  [
    "meta",
    {
      "content": "#ffffff",
      "name": "theme-color",
    },
  ],
  [
    "meta",
    {
      "content": "VuePress",
      "name": "application-name",
    },
  ],
  [
    "meta",
    {
      "content": "yes",
      "name": "mobile-web-app-capable",
    },
  ],
  [
    "link",
    {
      "crossorigin": "use-credentials",
      "href": "/manifest.webmanifest",
      "rel": "manifest",
    },
  ],
  [
    "link",
    {
      "href": "/img/icon/appleIcon152.png",
      "rel": "apple-touch-icon",
    },
  ],
  [
    "link",
    {
      "color": "#46bd87",
      "href": "/icons/safari-pinned-tab.svg",
      "rel": "mask-icon",
    },
  ],
]
`;

exports[`Test head function > should not generate tags if they exist 1`] = `
[
  [
    "link",
    {
      "href": "/icon.ico",
      "rel": "icon",
    },
  ],
  [
    "meta",
    {
      "content": "#ffffff",
      "name": "theme-color",
    },
  ],
  [
    "meta",
    {
      "content": "VuePress",
      "name": "application-name",
    },
  ],
  [
    "meta",
    {
      "content": "yes",
      "name": "mobile-web-app-capable",
    },
  ],
  [
    "link",
    {
      "crossorigin": "use-credentials",
      "href": "/manifest.webmanifest",
      "rel": "manifest",
    },
  ],
  [
    "link",
    {
      "href": "/img/icon/appleIcon152.png",
      "rel": "apple-touch-icon",
    },
  ],
  [
    "link",
    {
      "color": "#46bd87",
      "href": "/icons/safari-pinned-tab.svg",
      "rel": "mask-icon",
    },
  ],
]
`;

exports[`Test head function > should read manifest 1`] = `
[
  [
    "meta",
    {
      "content": "VuePress",
      "name": "application-name",
    },
  ],
  [
    "meta",
    {
      "content": "yes",
      "name": "mobile-web-app-capable",
    },
  ],
  [
    "meta",
    {
      "content": "#46bd87",
      "name": "theme-color",
    },
  ],
  [
    "link",
    {
      "crossorigin": "use-credentials",
      "href": "/manifest.webmanifest",
      "rel": "manifest",
    },
  ],
  [
    "link",
    {
      "href": "/assets/icon/chrome-192.png",
      "rel": "icon",
      "sizes": "192x192",
      "type": "image/png",
    },
  ],
  [
    "link",
    {
      "href": "/assets/icon/chrome-192.png",
      "rel": "apple-touch-icon",
    },
  ],
]
`;
